File: content/snippets/go/language_basics/functions_and_errors.go
package main

import (
	"errors"
	"fmt"
)

func add(a, b int) int {
	return a + b
}

func safeDiv(a, b float64) (float64, error) {
	if b == 0 {
		return 0, errors.New("division by zero")
	}
	return a / b, nil
}

func main() {
	fmt.Println("add(2, 3):", add(2, 3))

	if v, err := safeDiv(10, 2); err == nil {
		fmt.Println("safeDiv(10, 2):", v)
	}

	if _, err := safeDiv(10, 0); err != nil {
		fmt.Println("error:", err)
	}
}

File: content/snippets/cpp/language_basics/functions_and_errors.cpp
#include <iostream>
#include <optional>
#include <stdexcept>

int add(int a, int b) {
    return a + b;
}

std::optional<double> safe_div(double a, double b) {
    if (b == 0.0) {
        return std::nullopt;
    }
    return a / b;
}

int main() {
    std::cout << "add(2, 3): " << add(2, 3) << "\n";

    if (auto result = safe_div(10, 2)) {
        std::cout << "safe_div(10, 2): " << *result << "\n";
    }

    try {
        auto bad = safe_div(10, 0);
        if (!bad) {
            throw std::runtime_error("division by zero");
        }
    } catch (const std::exception& ex) {
        std::cerr << "error: " << ex.what() << "\n";
    }

    return 0;
}
